src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_QueryLambdaRewriter.vb

```diff
@@ -0,0 +1,83 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports Microsoft.CodeAnalysis.VisualBasic
+Imports Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor
+Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+Namespace Microsoft.CodeAnalysis.Semantics
+    Partial Friend NotInheritable Class VisualBasicOperationFactory
+        Private Shared Function RewriteQueryLambda(node As BoundQueryLambda) As BoundNode
+            ' We rewrite query lambda into regular lambda with 2 passes.
+            ' Pass 1 uses helper methods from LocalRewriter to do the lowering. This introduces large number of DAGs.
+            ' Pass 2 walks over the lowered tree and replaces duplicate bound nodes in the tree with their clones - this is a requirement for the Operation tree.
+            ' Note that the rewriter also rewrites all the query lambdas inside the body of this query lambda.
+
+            Try
+                Dim pass1Rewriter As New QueryLambdaRewriterPass1
+                Dim rewrittenLambda As BoundLambda = DirectCast(pass1Rewriter.VisitQueryLambda(node), BoundLambda)
+
+                Dim pass2Rewriter As New QueryLambdaRewriterPass2
+                Return pass2Rewriter.VisitLambda(rewrittenLambda)
+            Catch ex As CancelledByStackGuardException
+                Return node
```

Comment:
>Return node [](start = 16, length = 11)

Handling potential stack overflow in this manner can be a source of hard to repro bug reports. Let's open a follow up issue to confirm whether we want to use the guard (BoundTreeRewriterWithStackGuard) at all.

Reply:
Filed [https://github.com/dotnet/roslyn/issues/22193](https://github.com/dotnet/roslyn/issues/22193)

---
In reply to: [139535884](https://github.com/dotnet/roslyn/pull/21356#discussion_r139535884) [](ancestors = 139535884)